[DTensor][1/N] add forward layer norm support

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113244
* __->__ #113105

**Summary**:
This PR adds DTensor implementation for ATen op `native_layer_norm`.

**Test**:
`pytest test/distributed/_tensor/test_dtensor_ops.py -s -k layer_norm`


cc @wanchaol

XilunWu committed:
[DTensor][1/N] add forward layer norm support

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -251,3 +251,100 @@ def softmax_bwd_rule(op_schema: OpSchema) -> OutputSharding:
+def _replicate_dims_start_at(
+    placements: Sequence[Placement], starting_dim_optional: Optional[int] = 0
+) -> Tuple[Placement, ...]:
+    starting_dim = 0 if starting_dim_optional is None else starting_dim_optional
```
if you are doing this check, why not just make `starting_dim_optional: int = 0`?

XilunWu replied:
that's right. I keep mixing default arguments with optional type hints, in a wrong way....

wanchaol commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -251,3 +251,100 @@ def softmax_bwd_rule(op_schema: OpSchema) -> OutputSharding:
+) -> Tuple[Placement, ...]:
+    starting_dim = 0 if starting_dim_optional is None else starting_dim_optional
+    new_placements = []
+    for idx, placement in enumerate(placements):
```
I think this looks wrong, you are misusing `mesh` idx and the tensor dim idx. 

What we want is to make sure the tensor dimension starting from a certain dimension, there's no sharding, but it seems that here we are checking the mesh idx instead.

XilunWu replied:
i see. that's why the code in other places is using `dim_map`....

wanchaol commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -251,3 +251,100 @@ def softmax_bwd_rule(op_schema: OpSchema) -> OutputSharding:
+        else list(normalized_shape)
+    )
+
+    input_shape = input_schema.output_shape
```
this is not used?

wanchaol commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -251,3 +251,100 @@ def softmax_bwd_rule(op_schema: OpSchema) -> OutputSharding:
+            weight_src_spec = weight_schema.strategies[idx].output_spec
+            weight_target_spec = DTensorSpec(
+                mesh=mesh,
+                placements=_replicate_dims_start_at(weight_src_spec.placements, 0),
```
are we saying that we only support weight/bias replication for now? worth adding a comment

XilunWu replied:
Yes, to be more specific, we allow sharded weight/bias but we will replicate them before feeding them into the op.

wanchaol reviewed (commented):
first pass, please see the inline comments, I think the semantics of `replicate_dim_start_at` looks wrong and it might make all the tensor dims be replicated.

XilunWu commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -251,3 +251,100 @@ def softmax_bwd_rule(op_schema: OpSchema) -> OutputSharding:
+
+
+def layer_norm_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> OpStrategy:
+    # args must be: input, normalized_shape, weight, bias, eps
```
Good question. I did two things to testify:
1. `assert len(op_schema.args_schema) == 5` on line 258 to see if the schema is always a quintuple and the dtensor test has no error raised. 
2. However, this could be likely due to we always passed in `weight` and `bias` so I also added `assert bias is not None`. And the dtensor test raises error saying `bias` could be `None`. So we can tell that dtensor test actually tests the case where `bias` is optional.

Based on these two evidence, `weight`/`bias` always exist even if they're `None`.

XilunWu commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -251,3 +251,100 @@ def softmax_bwd_rule(op_schema: OpSchema) -> OutputSharding:
+        bias_schema,
+        _,
+    ) = op_schema.args_schema
+    # the current layer norm implementation requires that all input DTensor's sharding
```
Are strategy based op implementation mixed with ShardingProp based op implementation in sharding propagation? If so, I'll need to add compatible code to handle that.

wanchaol replied:
they are handled separately now, eventually we will get rid of the sharding prop rule based implementation.

XilunWu commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -251,3 +251,100 @@ def softmax_bwd_rule(op_schema: OpSchema) -> OutputSharding:
+    # must be in form of OpStrategy
+    assert isinstance(input_schema, OpStrategy)
+    assert isinstance(normalized_shape, (int, Sequence, torch.Size))
+    normalized_size = (
```
oops, the code was override in rebase...

XilunWu replied:
note: forward fix `normalize_to_torch_size` in #113244 to handle the `size` argument of type `int`, `Sequence`, or `torch.Size`. This fix also enables 3 dtensor op tests.

XilunWu committed:
Update on "[DTensor][1/N] add forward layer norm support"


**Summary**:
This PR adds DTensor implementation for ATen op `native_layer_norm`.

**Test**:
`pytest test/distributed/_tensor/test_dtensor_ops.py -s -k layer_norm`


[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -252,3 +253,105 @@ def softmax_bwd_rule(op_schema: OpSchema) -> OutputSharding:
+    return tuple(new_placements)
+
+
+register_op_strategy(aten.native_layer_norm.default, schema_info=RuntimeSchemaInfo())(
```
why not use make this a decorator for the layernorm strategy? It's the only one that uses this strategy

wanchaol commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -252,3 +253,105 @@ def softmax_bwd_rule(op_schema: OpSchema) -> OutputSharding:
+) -> Tuple[Placement, ...]:
+    new_placements: List[Placement] = []
+    for p in placements:
+        if p.is_partial() or (isinstance(p, Shard) and p.dim < start_dim):
```
hmmm shouldn't this be `p.dim >= start_dim` instead?

wanchaol commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -252,3 +253,105 @@ def softmax_bwd_rule(op_schema: OpSchema) -> OutputSharding:
+            assert isinstance(weight_strategy, OpStrategy)
+            weight_src_spec = weight_strategy.strategies[idx].output_spec
+
+            # for the weight tensor, we replicate it on all dims if necessary
```
can you add a TODO here so that we remember to see if we can keep weight/bias sharded instead of replicating on all dims?

XilunWu replied:
quick question: in short, what is the ideal way to support sharded weight/bias? My understanding is op decompose. Is that right?

wanchaol replied:
I think we almost likely won't need to make things sharded there, if we are doing it the normalized shape also need to allow sharding, we probably need some sort of decomposition there. Right now let's just have the complete picture documented.

wanchaol commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -1,5 +1,5 @@
 # Copyright (c) Meta Platforms, Inc. and affiliates
-from typing import cast, List, Optional, Tuple
+from typing import cast, List, Optional, Sequence, Tuple
```
In addition to the op level tests, I think we should add a test in `test_math_ops.py` to use `distribute_module` to partition a nn.LayerNorm, and mark input as sharded in `prepare_input_fn`, to make sure things work for the forward case when we apply this to nn.Module, this can be a follow up PR

wanchaol reviewed (commented):
Have a few comments inlined otherwise good to go!

XilunWu committed:
Update on "[DTensor][1/N] add forward layer norm support"


**Summary**:
This PR adds DTensor implementation for ATen op `native_layer_norm`.

**Test**:
`pytest test/distributed/_tensor/test_dtensor_ops.py -s -k layer_norm`


[ghstack-poisoned]

wanchaol reviewed (approved):
lgtm!
